content/lessons/3_data_types/data-types.rs

```diff
@@ -0,0 +1,65 @@
+#[derive(Clone, Copy, Debug)]
+struct Position(i32, i32); // tuple struct
+
+// Could Hero derive the Copy trait?
+#[derive(Clone, Debug)]
+struct Hero {
+    name: String,
+    level: u32,
+    experience: u32,
+    position: Position,
+}
+
+// we can add methods to structs using the 'impl' keyword
+impl Hero {
+    // static method
+    fn new(name: String) -> Hero {
+        Hero {
+            name,
+            level: 1,
+            experience: 0,
+            position: Position(0, 0),
+        }
+    }
+}
+
+// multiple impl blocks are possible for one struct
+impl Hero {
+    // instance method, first argument (self) is the calling instance
+    fn get_distance(&self, pos: Position) -> i32 {
+        // fields of tuples and tuple structs can be accessed through 'tuple.[i]'
+        (pos.0 - self.position.0).abs() + (pos.1 - self.position.1).abs()
+    }
+
+    // mutable borrow of self allows to change instance fields
+    fn level_up(&mut self) {
+        self.experience = 0;
+        self.level += 1;
+    }
+
+    // 'self' is not borrowed here and will be moved into the method
+    fn die(self) {
+        println!("Here lies {}, a hero who reached level {}. RIP.", self.name, self.level);
+    }
+}
+
+fn main() {
+    let mut hero: Hero = Hero::new(String::from("Marty The Brave"));
+    hero.level_up(); // 'self' is always passed implicitly
+
+    // fields other than 'name' will be the same as in 'hero'
+    let brother = Hero {
+        name: String::from("Steve The Normal Guy"),
+        ..hero.clone()
```

Comment:
Btw, idk if we want to introduce Eq in this module, if we do we could do:

```rust
let brother = Hero {
  name: String::from("Steve The Normal Guy"),
  ..hero
};

let mut twin = hero.clone();
assert_eq!(hero, twin);
twin.level_up();
assert_ne!(hero, twin);
hero.level_up();
assert_eq!(hero, twin);
```

and 

```rust
let position = Position(42, 0);
let other_position = position;
println!("Current position is {:?}", other_position);
let distance = hero.distance(position); // look ma, no clone!
// despite the fact that this function takes position by value, it is not moved, but copied, because the Position struct is Copy
assert_eq!(distance, 42);
```